File: 0049-groupAnagrams.js
// https://leetcode.com/problems/group-anagrams/

function groupAnagrams(strs) {
    const anagrams = {};
    const offset = 'a'.charCodeAt(0);
    
    for (const str of strs) {
        const key = createKey(str);
        
        if (anagrams[key] === undefined) {
            anagrams[key] = [];
        }
        anagrams[key].push(str);
    }
    
    function createKey(str) {
        const counts = new Array(26).fill(0);
        
        for (const char of str) {
            const i = char.charCodeAt(0) - offset;
            counts[i]++;
        }
        
        return counts.join(',');
    }
    
    return Object.values(anagrams);
}

File: java/0049-groupAnagrams.java
// https://leetcode.com/problems/group-anagrams/

class Solution {
    private static final int OFFSET = 'a';
    
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagrams = new HashMap<>();
        
        for (String str : strs) {
            String key = getKey(str);
            
            anagrams.putIfAbsent(key, new ArrayList<>());
            anagrams.get(key).add(str);
        }
        
        return new ArrayList<>(anagrams.values());
    }
    
    private String getKey(String str) {
        char[] counts = new char[26];
        
        for (char c : str.toCharArray()) {
            counts[c - OFFSET]++;
        }
        
        return String.valueOf(counts);
    }
}